### DISCUSSION TITLE
Clarity on npmrc packagerules

### DISCUSSION THREAD

**Discussion by HenrikPoulsen** [renovatebot/renovate#15054] [Request Help] [answered]
Hello.
I have been running with npmrc in packagerules for a specific package as we spoke about previously here #12312
Turns out that this does not work as intended.
I have been doing
```
{
      matchPackageNames: ['packageName'],
      npmrc:
        'registry=https://customregistry',
    },
```
And that did result in the package getting updates.
However it seems like this has caused all npm packages to now also use this custom registry, which is very much not what I would be expecting.
I have gone looking into the code and found this https://github.com/renovatebot/renovate/blob/a2b47c822f1cb35c8959636b20ced427c093f3ea/lib/modules/datasource/npm/npmrc.ts#L168
It only seems to look at `matchPackagePrefixes` and nothing else.
So I tried changing to use that field, but that does nothing, it's still undefined.
So then I looked at where the packageRules comes from and it goes in here for whatever reason https://github.com/renovatebot/renovate/blob/a2b47c822f1cb35c8959636b20ced427c093f3ea/lib/modules/datasource/npm/npmrc.ts#L94
and blankly applies it as the registry to use for the entire npm datasource.

Is this the intended result for this?
What I am trying to do is to just override the registry for one specific package, and let all other packages be as they were, using the default registry or the npmrc from the repo if there is one.
Is there any way to do this?

**Comment by rarkins** [comment, marked as answer]
It's an expected limitation that the `npmrc` is "shared" per repo. Usually scoped rules mean it doesn't conflict.

The npm datasources now supports customer registry urls: https://github.com/renovatebot/renovate/blob/e7899422fa86501fc9ef7adc810ce1c9956036f0/lib/modules/datasource/npm/index.ts#L13

Therefore try this instead:

```
{
      matchPackageNames: ['packageName'],
      registryUrls:
        ['https://customregistry'],
    },
```

**Reply by HenrikPoulsen** [reply]
Ah, great. Glad to see registryUrls work. I will use that. Thanks
